toy_2d/src/ilqr.py

```diff
@@ -204,18 +210,12 @@ def backward_pass(self,  xx, uu):
         """
         dd = [np.zeros((self.nu,))] * (self.N - 1)
         KK = [np.zeros((self.nu, self.nx))] * (self.N - 1)
-        # TODO: compute backward pass
-        V_xp1 = np.zeros((self.nx)) #jacobian of value func at the next state
-        V_xxp1 = np.zeros((self.nx, self.nx)) #hess of value func at the next state
-        for i in range(self.N-1,-1,-1):
-            # print(i)
+        final_state = xx[self.N-1]
+        V_xp1 = self.grad_terminal_cost(final_state) #jacobian of value func at the next state
+        V_xxp1 = self.hess_terminal_cost(final_state) #hess of value func at the next state
+        for i in range(self.N-2,-1,-1):
```

Comment:
Why did this change from starting at `N-1` to now `N-2`?  I think it should probably still be `N-1` unless I'm missing something.

Reply:
So the state vector has a size N (0,N-1) and the controls vector has a size of N-1 (0,N-2) as seen in the assertion check I have put during the start of the function. No action can be applied at the last state because ideally that would be the goal state. Since I removed the if condition which was getting triggered at N-1 value of i, I have to change this to N-2 otherwise uu[i] would throw and out of bounds error.

Reply:
Ah, I missed the `continue` in your old code that made the loop move on to `N-2`.  Makes sense, thanks for clarifying.